### PR TITLE
Bump rocksdb pointer.

### PR DESCRIPTION
**Pull request by petermattis**

### CODE REVIEW

**Comment by bdarnell** [comment]
This circleci failure is one reason why I do upstream merges without review; you can't run CI on the main repo until the submodule is updated. LGTM assuming things work on circleci after you've merged the other branch. Are there any changes of note in rocksdb 3.9?

**Comment by petermattis** [comment]
Hmm, I'm seeing the same error with `make acceptance`. Strange because the SHA it is complaining about should definitely be present. Is docker caching the rocksdb repo somehow?

**Comment by bdarnell** [comment]
Are you sure it's using the right commit? I don't see 9cda7cb in cockroachdb/rocksdb. The latest commit there is 853b545 (and your earlier merge commit was ef4f2af)

**Comment by petermattis** [comment]
9cda7cb corresponds to the tag "rocksdb-3.9". Looks like it is there to me: https://github.com/cockroachdb/rocksdb/commit/9cda7cb77b0c7208a63579c7e79252f23db92f67

**Comment by bdarnell** [comment]
Weird. That didn't come down with `git fetch` for me. Normally I'd just say it's because tags behave strangely and often need to be pushed separately, but I did get several other tags in this fetch:
```
$ git fetch
remote: Counting objects: 6795, done.
remote: Compressing objects: 100% (2827/2827), done.
remote: Total 6795 (delta 4879), reused 5423 (delta 3800)
Receiving objects: 100% (6795/6795), 2.86 MiB | 3.81 MiB/s, done.
Resolving deltas: 100% (4879/4879), done.
From https://github.com/cockroachdb/rocksdb
   ed3aa7b..853b545  master     -> origin/master
 * [new tag]         2.8.fb     -> 2.8.fb
 * [new tag]         rocksdb-3.1 -> rocksdb-3.1
 * [new tag]         rocksdb-3.2 -> rocksdb-3.2
 * [new tag]         rocksdb-3.3 -> rocksdb-3.3
 * [new tag]         rocksdb-3.7 -> rocksdb-3.7
 * [new tag]         rocksdb-3.8 -> rocksdb-3.8
```

In any case, the submodule needs to be set to the merge commit that combines our fork with upstream, not any of the upstream commits.

**Comment by petermattis** [comment]
Our local changes are no longer necessary (similar changes were made upstream). Should I change our fork's master branch to point to the rocksdb-3.9 tag?

**Comment by petermattis** [comment]
Switching to set the submodule commit to our rocksdb fork's master branch seems to have fixed the circleci problem.

**Comment by bdarnell** [comment]
Our fork's master branch shouldn't really matter, although it wouldn't hurt to replace it with upstream. (`git checkout rocksdb-3.9; git push -f origin HEAD:master`).

**Comment by petermattis** [comment]
Ugh, now it's failing because one of the docker layers has cached out of date rocksdb build information (namely that db/log_writer.o depends on a non-existent header file).

**Comment by tschottdorf** [comment]
You can just wipe the cache on CircleCI. Updating the submodules is a bit awkward at times because we've hidden the RocksDB build early in the process for performance.
I'm retriggering your build now and cleaning the cache.

**Comment by petermattis** [comment]
I'm pretty sure my change to the docker build process will fix this. See the latest commit on this PR.

**Comment by tschottdorf** [comment]
I don't think you want to use `git submodule status` for scripting, this one looks "righter": http://stackoverflow.com/questions/3983829/how-can-i-get-a-git-submodules-associated-commit-id-from-a-past-commit-in-the-p

**Comment by petermattis** [comment]
I could switch to using `git ls-tree HEAD _vendor/rocksdb`, but that gives the same output as `git submodule status`. Is there something problematic about `git submodule status` that I'm not seeing?

**Comment by tschottdorf** [comment]
it's a porcelain command, and your test fails because (?) the command output is parsed incorrectly (commit it prefixed with a "-" when module is behind or something). Probably better to use a lower-level command that just displays the SHA. I'm a bit confused about the initial test failure. Where did the submodule update go wrong in in #417?

**Comment by petermattis** [comment]
The problem in #417 was that the rocksdb "base" layer was made with rocksdb build information from August and various files have moved and been renamed since then. In particular, `port/atomic_pointer.h` no longer exists but was still be referenced by the cached build information. We really want to do a `make clean` in rocksdb whenever the SHA is changed. Rather than doing this, my change is invalidating re-cloning rocksdb whenever the SHA changes which then causes the rocksdb build to be completely clean.

Strange that `make acceptance` is working locally for me. I'll switch to using `git ls-tree`.

**Comment by tschottdorf** [comment]
Strange that make acceptance is working locally for me. I'll switch to using git ls-tree.
`git submodule status` has different prefixes for the SHA depending on what's present, probably that's why. When your test passes, can you make sure it doesn't completely invalidate build caching? Simply retriggering the build and making sure it caches all the steps as before (or just making sure it takes about half as long as the first time around).

**Comment by petermattis** [comment]
Yeah, just read that same bit of documentation myself. `git ls-tree` seemed to have worked fine. PTAL.

**Comment by petermattis** [comment]
`make acceptance` is reusing the caches. I'll also verify that circleci is.

File: build/devbase/Dockerfile.template [hunk 1]
```diff
@@ -30,21 +30,23 @@ RUN mkdir -p $ROACHPATH && \
...
+# update" here even though we'll run it again later.
+RUN \
+ cd $ROCKSDBPATH && git clone https://github.com/cockroachdb/rocksdb.git && \
+ cd $ROCKSDBPATH/rocksdb && git checkout ROCKSDB_SHA && \
+ cd $COREOSPATH && git clone https://github.com/cockroachdb/etcd.git && \
+ cd $COREOSPATH/etcd && git checkout ETCD_SHA
+
+# Avoid adding the Dockerfile which gets "touched" on every build and
+# would this invalidate subsequent layer caching.
+ADD *.sh /cockroach/build/devbase/
```
Comment (tschottdorf) [inline build/devbase/Dockerfile.template]: why is this and the next one here?

Comment (tschottdorf) [inline build/devbase/Dockerfile.template]: nvm, I see it.

**Comment by tschottdorf** [comment]
LGTM, assuming the caching stuff works out as it probably will.

**Comment by petermattis** [comment]
Caching is working for both `make acceptance` and circleci. I don't think the `git submodule update` in the top-level Dockerfile is necessary any more, but I'll leave that change to a subsequent PR.

**Pull request merged**